core/impl/src/main/java/com/blazebit/persistence/impl/AbstractCTECriteriaBuilder.java

```diff
@@ -201,38 +209,83 @@ public CTEInfo createCTEInfo() {
+
+            if (jpaMetamodelAccessor.isJoinable(lastAttribute) || lastAttribute.getPersistentAttributeType() == Attribute.PersistentAttributeType.EMBEDDED) {
+                // We have to map *-to-one relationships to their id or unique props
                 // NOTE: Since we are talking about *-to-ones, the expression can only be a path to an object
                 // so it is safe to just append the id to the path
-                Expression selectExpression = selectManager.getSelectInfos().get(bindingEntry.getValue()).getExpression();
+                Expression selectExpression = selectManager.getSelectInfos().get(tupleIndex).getExpression();
 
                 // TODO: Maybe also allow Treat, Case-When, Array?
                 if (selectExpression instanceof NullExpression) {
                     // When binding null, we don't have to adapt anything
                 } else if (selectExpression instanceof PathExpression) {
-                    PathExpression pathExpression = (PathExpression) selectExpression;
-                    // Only append the id if it's not already there
-                    if (!idAttribute.getName().equals(pathExpression.getExpressions().get(pathExpression.getExpressions().size() - 1).toString())) {
-                        pathExpression.getExpressions().add(new PropertyExpression(idAttribute.getName()));
+                    boolean firstBinding = true;
+                    final Collection<String> embeddedPropertyNames;
+
+                    if (lastAttribute.getPersistentAttributeType() == Attribute.PersistentAttributeType.EMBEDDED) {
+                        embeddedPropertyNames = new TreeSet<>(JpaMetamodelUtils.getEmbeddedPropertyNames((EmbeddableType<?>) ((SingularAttribute<?, ?>) lastAttribute).getType()));
+                    } else {
+                        embeddedPropertyNames = new TreeSet<>(cbf.getJpaProvider().getIdentifierOrUniqueKeyEmbeddedPropertyNames(cteType, attributeName));
+                    }
+
+                    PathExpression baseExpression = embeddedPropertyNames.size() > 1 ?
+                            ((PathExpression) selectExpression).clone(false) : ((PathExpression) selectExpression);
+
+                    for (String embeddedPropertyName : embeddedPropertyNames) {
+                        PathExpression pathExpression = firstBinding ?
+                                ((PathExpression) selectExpression) : baseExpression.clone(false);
+
+                        // TODO: this is supposed to check whether someAssociation is bound to someAlias.id
+                        // Actually, the id names can collide in the case of IdClass or EmbeddedIds, so a true check
+                        // should check whether the type of the path reference of the path expression is a basic type.
+                        // However, this seems not possible at this time because for some tests pathExpression.getPathReference()
+                        // returns null.
```

Comment:
@beikov Can you have a look at this comment?

Reply:
That uses the same logic unfortunately. I'll leave the check for what it is for now.

```
java.lang.NullPointerException
	at com.blazebit.persistence.impl.JpaUtils.getAttributeForJoining(JpaUtils.java:39)
	at com.blazebit.persistence.impl.AbstractCTECriteriaBuilder.prepareAndGetAttributes(AbstractCTECriteriaBuilder.java:247)
	at com.blazebit.persistence.impl.RecursiveCTECriteriaBuilderImpl.createCTEInfo(RecursiveCTECriteriaBuilderImpl.java:94)
	at com.blazebit.persistence.impl.CTEManager.onBuilderEnded(CTEManager.java:183)
	at com.blazebit.persistence.impl.RecursiveCTECriteriaBuilderImpl.onBuilderEnded(RecursiveCTECriteriaBuilderImpl.java:82)
	at com.blazebit.persistence.impl.SelectCTECriteriaBuilderImpl.end(SelectCTECriteriaBuilderImpl.java:47)
```

Reply:
In that case you would have to call `implicitJoin(expression, true, null, ClauseType.SELECT, null, false, false, false, false)` before that.

Reply:
```java
                        joinManager.implicitJoin(pathExpression, true, null, ClauseType.SELECT, null, false, false, false, false);
                        AttributeHolder attributeForJoining = JpaUtils.getAttributeForJoining(getMetamodel(), pathExpression);

                        if (attributeForJoining.getAttributeType().getPersistenceType().equals(Type.PersistenceType.BASIC)) {
                            throw new IllegalStateException("An association should be bound to its association type and not its identifier type");
                        }
```

For example, fails on some occasions with:

```
java.lang.IllegalArgumentException: Attribute 'a' not found on type 'Issue571Test$MyEntity'
	at com.blazebit.persistence.parser.PathTargetResolvingExpressionVisitor.visit(PathTargetResolvingExpressionVisitor.java:224)
	at com.blazebit.persistence.parser.expression.PropertyExpression.accept(PropertyExpression.java:41)
	at com.blazebit.persistence.parser.PathTargetResolvingExpressionVisitor.visit(PathTargetResolvingExpressionVisitor.java:316)
	at com.blazebit.persistence.parser.expression.PathExpression.accept(PathExpression.java:74)
	at com.blazebit.persistence.impl.JpaUtils.getAttributeForJoining(JpaUtils.java:66)
	at com.blazebit.persistence.impl.JpaUtils.getAttributeForJoining(JpaUtils.java:57)
	at com.blazebit.persistence.impl.AbstractCTECriteriaBuilder.prepareAndGetAttributes(AbstractCTECriteriaBuilder.java:249)
	at com.blazebit.persistence.impl.RecursiveCTECriteriaBuilderImpl.createCTEInfo(RecursiveCTECriteriaBuilderImpl.java:94)
	at com.blazebit.persistence.impl.CTEManager.onBuilderEnded(CTEManager.java:183)
	at com.blazebit.persistence.impl.RecursiveCTECriteriaBuilderImpl.onBuilderEnded(RecursiveCTECriteriaBuilderImpl.java:82)
	at com.blazebit.persistence.impl.SelectCTECriteriaBuilderImpl.end(SelectCTECriteriaBuilderImpl.java:47)
	at com.blazebit.persistence.testsuite.Issue571Test$2.work(Issue571Test.java:94)
	at com.blazebit.persistence.testsuite.tx.TxSupport.transactional(TxSupport.java:33)
	at com.blazebit.persistence.testsuite.AbstractCoreTest.transactional(AbstractCoreTest.java:383)
	at com.blazebit.persistence.testsuite.Issue571Test.testBindingCteAssociationToEntity(Issue571Test.java:81)
```

I.e. it seems that I have to resolve some aliases manually? I'm honestly not sure whether we can *easily* strengthen the check, and I think the existing check will be sufficient 99% of the time. (I hardly imagine anyone every having a domain with a nested identifier with a poor naming scheme as `entity.id.id.id` where `id`'s are either associations or embeddables).

Reply:
Nevermind, 


```java
                       joinManager.implicitJoin(pathExpression, true, null, ClauseType.SELECT, null, false, false, false, false);
                        
                        if (pathExpression.getPathReference().getType().getPersistenceType().equals(Type.PersistenceType.BASIC)) {
                            throw new IllegalStateException("An association should be bound to its association type and not its identifier type");
                        }
```

Seems to do the trick.